src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -1202,7 +1242,27 @@ impl LinkCollector<'_, '_> {
             }
         }
 
-        let res = self.resolve_with_disambiguator(&key, diag.clone()).and_then(|(res, def_id)| {
+        let mut candidates = self.resolve_with_disambiguator(&key, diag.clone());
+
+        // FIXME: it would be nice to check that the feature gate was enabled in the original crate, not just ignore it altogether.
+        // However I'm not sure how to check that across crates.
+        if let Some(candidate) = candidates.get(0) &&
+            candidate.0 == Res::Primitive(PrimitiveType::RawPointer) &&
+            key.path_str.contains("::") // We only want to check this if this is an associated item.
+        {
+            if key.item_id.is_local() && !self.cx.tcx.features().intra_doc_pointers {
+                self.report_rawptr_assoc_feature_gate(diag.dox, &diag.link_range, diag.item);
+                return None;
+            } else {
+                candidates = vec![candidates[0]];
+            }
+        }
+
+        if candidates.len() > 1 && !ambiguity_error(self.cx, &diag, &key.path_str, &candidates) {
+            candidates = vec![candidates[0]];
+        }
```

Comment:
I don't understand what's going on here. Why are you arbitrarily picking a candidate here? This happens too late, you can no longer be sure that all the candidates came from trait selection instead of from somewhere where you need to report the ambiguity.

Reply:
It allows to fix the following case (using the code example you provided):

```rust
//! [usize::Item]

pub trait Foo { type Item; }
pub trait Bar { type Item; }

impl Foo for usize { type Item = u32; }
impl Bar for usize { type Item = i32; }
```

Without the deduplication logic present in `ambiguity_error`, it would output:

```
error: `usize::Item` is both an associated type and an associated type
 --> tests/rustdoc-ui/intra-doc/issue-108653-associated-items-9.rs:5:6
  |
5 | //! [usize::Item]
  |      ^^^^^^^^^^^ ambiguous link
  |
note: the lint level is defined here
 --> tests/rustdoc-ui/intra-doc/issue-108653-associated-items-9.rs:3:9
  |
3 | #![deny(warnings)]
  |         ^^^^^^^^
  = note: `#[deny(rustdoc::broken_intra_doc_links)]` implied by `#[deny(warnings)]`
help: to link to the associated type, prefix with `type@`
  |
5 | //! [type@usize::Item]
  |      +++++
help: to link to the associated type, prefix with `type@`
  |
5 | //! [type@usize::Item]
  |      +++++

error: aborting due to previous error
```

I'll add a comment to explain why this is needed.